### PR TITLE
JBPM-6894 - User should be able to define case, process, activity lev…

### PR DESCRIPTION
**Pull request by mswiderski**
…el SLA

depends on https://github.com/kiegroup/droolsjbpm-integration/pull/1378

### CODE REVIEW

**Comment by mswiderski** [comment]
anyone willing to review?

**Comment by tsurdilo** [comment]
@mswiderski looks good to me +1 !

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    private static final String SPAN_NAME = "sla-compliance";
+    private static final String SLA_STYLE = "label";
+    private static final String SLA_NA_PRESENT_STYLE = "label-default";
+    private static final String SLA_PENDING_PRESENT_STYLE = "label-primary";
+    private static final String SLA_MET_PRESENT_STYLE = "label-success";
+    private static final String SLA_ABORTED_PRESENT_STYLE = "label-warning";
+    private static final String SLA_VIOLATED_PRESENT_STYLE = "label-danger";
+
+    
+    private List<String> descriptions;
```
Comment (cristianonicolai) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: minor adjustment in spacing between elements

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import org.jboss.errai.ui.shared.api.annotations.DataField;
+import org.jboss.errai.ui.shared.api.annotations.Templated;
+import org.kie.api.runtime.process.ProcessInstance;
+
+import com.google.gwt.cell.client.AbstractCell;
+import com.google.gwt.safehtml.shared.SafeHtmlBuilder;
+
+@Dependent
+@Templated(value = "SLAComplianceCell.html")
+public class SLAComplianceCell extends AbstractCell<Integer> implements IsElement {
```
Comment (cristianonicolai) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: @mswiderski we're moving to the new elemental API for new UI screens, I'm making some comments in here so it can be adjusted to the new API.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+package org.jbpm.workbench.common.client.util;
+
+import java.util.List;
+
+import javax.enterprise.context.Dependent;
+import javax.inject.Inject;
+
+import org.jboss.errai.common.client.dom.DOMTokenList;
+import org.jboss.errai.common.client.dom.Span;
+import org.jboss.errai.ui.client.local.api.IsElement;
```
Comment (cristianonicolai) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: should be replaced by _org.jboss.errai.common.client.api.elemental2.IsElement_

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    private static final String SLA_PENDING_PRESENT_STYLE = "label-primary";
+    private static final String SLA_MET_PRESENT_STYLE = "label-success";
+    private static final String SLA_ABORTED_PRESENT_STYLE = "label-warning";
+    private static final String SLA_VIOLATED_PRESENT_STYLE = "label-danger";
+
+    
+    private List<String> descriptions;
+    @Inject
+    @DataField(SPAN_NAME)
+    private Span content;
```
Comment (cristianonicolai) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: To adjust this element to the new API you could use:
`@Inject
    @DataField(SPAN_NAME)
@Named("span")
    private HTMLElement content;`

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                       SafeHtmlBuilder sb) {
+
+        DOMTokenList tooltipClasses = content.getClassList();
+        tooltipClasses.add(SLA_STYLE);
+        String description = descriptions.get(0);
+        switch (value) {
+            case ProcessInstance.SLA_NA:
+                description = descriptions.get(1);
+                tooltipClasses.add(SLA_NA_PRESENT_STYLE);
+                tooltipClasses.remove(SLA_PENDING_PRESENT_STYLE);
```
Comment (cristianonicolai) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: Perhaps is not necessary to remove all unrelated css classes as this element isnt reused? If so, it could be done at the start, removing all classes.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/details/ProcessInstanceDetailsPresenter.java [hunk 6]
```diff
@@ -167,5 +190,7 @@ public void callback(final List<NodeInstanceSummary> details) {
         HTML getCorrelationKeyText();
 
         HTML getParentProcessInstanceIdText();
+        
+        HTML getSlaComplianceText();
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/details/ProcessInstanceDetailsPresenter.java]: Returning the html element is a practice we're trying to get away, please update to only set the relevant text, something like _setSlaCompliance(String text)_

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/details/ProcessInstanceDetailsViewImpl.java [hunk 7]
```diff
@@ -43,6 +43,10 @@
     @Inject
     @DataField
     public HTML processVersionText;
+    
+    @Inject
+    @DataField
+    public HTML slaComplianceText;
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/details/ProcessInstanceDetailsViewImpl.java]: These text fields need fixing as they don't match the html element from the template (HTML vs Paragraph). If you could try to use _HTMLParagraphElement_ that would get us on the right path.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/resources/i18n/Constants.java [hunk 8]
```diff
@@ -131,7 +131,7 @@
 
     String Process_Instance_Details();
 
-    String Process_Definition_Version();
+    String Process_Definition_Version();
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/resources/i18n/Constants.java]: perhaps revert this change.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/details/ProcessInstanceDetailsPresenter.java [hunk 9]
```diff
@@ -118,6 +118,29 @@ public void callback(final ProcessInstanceSummary process) {
                     default:
                         break;
                 }
+                view.getStateText().setText(statusStr);
+                
+                String slaComplianceStr = constants.Unknown();
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/details/ProcessInstanceDetailsPresenter.java]: Would it be work moving this logic into a separated method that could also be easily tested?

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessRuntimeDataServiceImpl.java [hunk 10]
```diff
@@ -206,6 +206,8 @@ protected ProcessInstanceSummary build(ProcessInstance processInstance) {
                 processInstance.getCorrelationKey(),
                 processInstance.getParentId(),
                 null,
+                processInstance.getSlaCompliance(),
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-backend/src/main/java/org/jbpm/workbench/pr/backend/server/RemoteProcessRuntimeDataServiceImpl.java]: Would be nice to add a test that ensures the mapping between ProcessInstance and ProcessInstanceSummary works as expected.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+                tooltipClasses.add(SLA_VIOLATED_PRESENT_STYLE);               
+                break;
+            default:
+                description = descriptions.get(0);
+                tooltipClasses.add(SLA_NA_PRESENT_STYLE);
+                break;
+
+        }        
+        content.textContent = description; 
+        sb.appendHtmlConstant(content.innerHTML);
```
Comment (mswiderski) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: @cristianonicolai @mbarkley how to render complete HTMLElement here? There is no outerHtml thingy in elemental2.... so with this refactored code there are no coloured labels any more which pretty much makes this custom cell useless...

**Comment by mswiderski** [comment]
@cristianonicolai I made the changes you requested but so far I didn't find a way to achieve what was there before. Hints more than welcome.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java (hunk 11 above)
Comment (cristianonicolai) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: @mswiderski seems like outerHTML is actually missing in the API: https://github.com/google/elemental2/issues/37

I think in this case we could simplify things and just wrap with another element, it also seems that using Errai template would be unnecessary and actually share the element across multiple render calls, have a look in my suggestions here: https://gist.github.com/cristianonicolai/59d6411865dd1b1610cc2f66a315b2eb

Comment (mswiderski) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/util/SLAComplianceCell.java]: @cristianonicolai thanks a lot, this works perfectly. all good I believe, do you mind to take a look again?

**Comment by cristianonicolai** [comment]
looks good, thanks @mswiderski !
